File: C++/0203-remove-linked-list-elements.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
 public:
  ListNode* removeElements(ListNode* head, int val) {
    ListNode* temp;
    ListNode* prehead = new ListNode(0);
    ListNode* prev;
    prehead->next = head;
    prev = prehead;
    temp = head;
    while (temp != NULL) {
      if (temp->val == val) {
        prev->next = temp->next;
      } else {
        prev = temp;
      }
      temp = temp->next;
    }
    return prehead->next;
  }
};

File: 0203-remove-linked-list-elements/0203-remove-linked-list-elements.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        prev = dummy
        curr = head
        while curr:
            if curr.val == val:
                prev.next = curr.next
                curr = curr.next
            else:
                prev = curr
                curr = curr.next
        return dummy.next


head = ListNode()
val = 7
print(Solution().removeElements(head, val))

head = ListNode(1)
head.next = ListNode(2)
head.next.next = ListNode(6)
head.next.next.next = ListNode(3)
head.next.next.next.next = ListNode(4)
head.next.next.next.next.next = ListNode(5)
head.next.next.next.next.next.next = ListNode(6)
val = 6
print(Solution().removeElements(head, val))

head = ListNode(7)
head.next = ListNode(7)
head.next.next = ListNode(7)
head.next.next.next = ListNode(7)
val = 7
print(Solution().removeElements(head, val))
